Fix: Adds autofixer for no-plusplus rule. (fixes: #7505)

eelyafi opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[X ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (http://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**
This PR will add an autofixer to no-plusplus rule. Autofixer will skip `return i++;` since we can't safely fix this case.

**Is there anything you'd like reviewers to focus on?**
Not really.

not-an-aardvark commented on lib/rules/no-plusplus.js:
```diff
@@ -46,12 +46,28 @@ module.exports = {
                         operator: node.operator
-                    }
+                    },
+                    fix: fixerFunc ? fixerFunc : function() {}
```
Nit: this ternary expression seems redundant. Can you replace it with:

```js
fix: fixerFunc || function() {}
```

not-an-aardvark commented on tests/lib/rules/no-plusplus.js:
```diff
@@ -26,14 +26,31 @@ ruleTester.run("no-plusplus", rule, {
+            output: "for (i = 0; i < l; i++) { v += 1; }",
+            options: [{allowForLoopAfterthoughts: true}],
+            errors: [{ message: "Unary operator '++' used.", type: "UpdateExpression"}]
+        }
```
Can you add tests for cases like `++i` and `--i` as well? I think these can be fixed in all cases:

```js
foo = ++bar;
baz = --qux;

// is equivalent to:

foo = bar += 1;
baz = qux -= 1;
```

not-an-aardvark commented on lib/rules/no-plusplus.js:
```diff
@@ -46,12 +46,28 @@ module.exports = {
+
+                let fixerFunc;
+
+                if (node.parent.type !== "ReturnStatement") {
```
I don't think this check is sufficient. It's possible to use the result of `++` and `--` outside of a return statement:

```js
bar = 0;
foo = bar++;
// now foo = 0

bar = 0;
foo = bar += 1;
// now foo = 1; the fix changed the behavior of the code
```

Instead, I would recommend only doing a fix when the parent is an `ExpressionStatement`:

```js
if (node.parent.type === "ExpressionStatement" || node.prefix) {
  // do a fix
}
```

eelyafi replied:
should we fix for "ForStatement" as well ? otherwise this case won't be fixed:
`for (i = 0; i < l; i++) { console.log(i); }`

not-an-aardvark commented on tests/lib/rules/no-plusplus.js:
```diff
@@ -26,14 +26,31 @@ ruleTester.run("no-plusplus", rule, {
-        { code: "for (i = 0; i < l; i++) { console.log(i); }", errors: [{ message: "Unary operator '++' used.", type: "UpdateExpression"}] },
+        {
+            code: "function foo() { return i++; }",
+            errors: [{ message: "Unary operator '++' used.", type: "UpdateExpression"}]
```
Can you add an `output` property to this test? You can make it the same as the code:

```js
{
    code: "function foo() { return i++; }",
    output: "function foo() { return i++; }",
    errors: // ...
}
```

This will ensure that no fix is performed in this case.

not-an-aardvark commented on lib/rules/no-plusplus.js:
```diff
@@ -46,12 +46,28 @@ module.exports = {
+                        operator = " += 1";
+                    }
+
+                    const fixedCode = node.argument.name + operator;
```
I think using `node.argument.name` only works if `node.argument` is an `Identifier`; it will fail in other cases:

```js
foo.bar++;

// gets fixed to:

undefined += 1; // oops
```

I would recommend using `sourceCode.getText(node.argument)` instead of `node.argument.name`.

nzakas commented:
I'm not sure we should have autofixing for this. This is one of those situations where we can't really know what the correct fix (in general, we don't autofix rules beginning with `no-` because they only prevent you from doing something without prescribing a specific fix).

eelyafi commented:
@nzakas, thanks for the feedback. I think it's clear how the fix should look like. Also, as an eslint user, I think it's better to have an autofixer that will cover most of cases than nothing.

platinumazure commented:
@eelyafi The overriding concern here is we *do not* want to auto-fix code in such a way that it makes things worse. That could mean the resulting code is unparseable, or it could mean comments are removed or mangled, or it could mean that the behavior of the code has changed.

As a result, we are very cautious about adding auto-fixers that might change the behavior of code.

eelyafi commented:
@platinumazure can you please provide an example how changing `i++;` to `i += 1;` can lead to the results listed above ? It seems like except for `return i++;` it should work fine, isn't it ?

not-an-aardvark commented:
@eelyafi 

Before autofixing:

```js
bar = 0;
foo = bar++;
assert.strictEqual(foo, 0);
```

After autofixing:

```js
bar = 0;
foo = bar += 1;
asserti.strictEqual(foo, 0); // this will throw an error, because foo is 1 instead of 0
```

eelyafi commented:
@not-an-aardvark make sense, how to avoid this situation ? will this check fix this ?
```
if (node.parent.type === "ExpressionStatement" || node.prefix) {
  // do a fix
}
```

not-an-aardvark commented:
Yes, I think that will work.

not-an-aardvark commented on lib/rules/no-plusplus.js:
```diff
@@ -40,18 +40,36 @@ module.exports = {
+
+                let fixerFunc;
+
+                if (node.parent.type === "ExpressionStatement" || node.parent.type === "ForStatement" || node.prefix) {
```
I think this check isn't strict enough; it allows things like this:

```js
for (var foo = bar++; bar; baz) {} // `foo` will start out with the wrong value
```

To fix this, I would recommend changing the check a bit so that `++` expressions are only allowed in the "update" of the `for` loop (the last clause):

```js
if (
    node.parent.type === "ExpressionStatement" ||
    node.prefix ||
    (node.parent.type === "ForStatement" && node === node.parent.update)
) {
  // do a fix
}
```

nzakas commented:
I'd like to get a few more eyes on this before accepting. I'm still not convinced we should be autofixing this rule, but I'll defer to consensus if people feel differently.

mysticatea commented:
:+1: for this fixing.

But we should be careful to fix it since this changes the node type from `UpdateExpression` to `AssignmentExpression`. This implies that it changes operator precedence.

```js
let i = 9
console.log(++i === 10) // → true
```

```js
let i = 9
console.log(i += 1 === 10) // → 9
```

not-an-aardvark commented:
What about `node.prefix && astUtils.isParenthesized(node)`?

mysticatea commented:
@not-an-aardvark It works for me. But I think that each people has preference about assignments in the middle of expressions.

eelyafi commented:
@mysticatea, @not-an-aardvark I'm going to change the if statement to:
```
if (
  node.parent.type === "ExpressionStatement" ||
  (node.prefix && astUtils.isParenthesized(node)) ||
  (node.parent.type === "ForStatement" && node === node.parent.update)
) {
  // do fix
}
```

what cases will be covered ? I'm going to add this to the docs. Thanks

not-an-aardvark commented:
@eelyafi This covers cases like this:

```js
let i = 9;
console.log(++i === 10); // → should not get fixed

let i = 9;
console.log((++i) === 10); // → should get fixed
```

Would you mind adding tests for these cases as well? It would be good to make sure that the fixer is working correctly in these cases.

nzakas commented:
TSC Summary: This PR proposes to autofix the `no-plus-plus` rule by replacing `i++` with `i += 1` and `i--` with `i -= 1`. My feeling is that we shouldn't be doing this autofix do the locations where `i++` can be used and the fact that we don't know 100% what the preferred replacement for everyone should be. The counterargument is that we can do autofixes in some places with reasonable accuracy and `+=` and `-=` might be okay for most developers.

TSC Question: Do we want to autofix this rule?

eelyafi commented:
@nzakas we can add options to this rule in order to specify the preferred replacement. 

@not-an-aardvark for some reason tests are passing on my laptop, but failing on continuous integration. There are two errors:
`TypeError: astUtils.isParenthesized is not a function`
and 
`3) no-plusplus invalid let i = 9; console.log(++i === 10);:
      AssertionError: A fatal parsing error occurred: Parsing error: Unexpected token i`
do you know, by any chance, what may cause this inconsistency ?

not-an-aardvark commented:
@eelyafi We generally don't add rule options indicating how rules should be fixed; instead we just avoid fixing those cases.

> TypeError: astUtils.isParenthesized is not a function

Oops, I think the function is `astUtils.isParenthesised` (with an `s` instead of a `z`)

> Parsing error: Unexpected token i

This happens because `let` is an ES6 keyword, and the parser only checks ES5 by default. To fix this, you can either use `parserOptions: {ecmaVersion: 6}` on the test object, or change `let` to `var`.

```js
code: "var i = 9; console.log(++i === 10);"
```

> for some reason tests are passing on my laptop

That's a bit surprising; are you just using `npm test` to run the tests?

eelyafi commented:
@not-an-aardvark, it was my fault, I modified Makefile.js in order to run tests faster which resulted in passing failing tests. I updated the last commit. Now it should work fine. Thanks for the help !

eelyafi commented:
@nzakas, I just want to share my experience of using an autofixer like this one. I just used it on the Pinterest's codebase. It successfully fixed all the issues and only about 20 files left. It's a huge difference, it's much easier to fix 20 files by hands than hundreds. That's why, as an enduser, I would prefer to have an autofixer that will fix most of the cases than nothing..

eelyafi commented:
@platinumazure, @mysticatea do you mind to take a look at this PR ?

btmills commented:
After discussion in the 2016-11-10 TSC meeting, the team decided against including autofix for `no-plus-plus`. @eelyafi thank you for the hard work you put into this.

eelyafi commented:
@btmills sad and I don't understand why... as I mentioned above, it was super helpful..

alberto commented:
we don't think it's a 100% clear and risk free fix so we decided against it at this moment. you can still publish your own version of the rule, though

platinumazure commented:
@eelyafi We have to consider the ecosystem, including users who might have particular (manual) fixes in mind which the autofixer might not know about. If you find this useful, you can definitely feel free to implement a fixable version of `no-plusplus` as a custom rule or in a plugin, so that it can still be used by those who would find it useful. All we're saying here is that we aren't sure we want this in ESLint Core.

ilyavolodin commented:
@eelyafi Basically, the team felt that until we have a way for users to specify which rules they want autofixed, it's too presumptuous to assume that we know what users meant when they said don't use `++/--`. We do appreciate the work you put into this, it's just because there are so many people using ESLint, we tend to take the cautious approach.
You can read TSC meeting notes here: https://gitter.im/eslint/tsc-meetings/archives/2016/11/10

eelyafi commented:
@alberto, @platinumazure, @ilyavolodin, thanks for the response. That's awesome that Eslint cares so much about having highly reliable and safe rules/autofixers. I have some feedback from my side as an end user of Eslint. It looks like adding a new feature takes too much effort because it should be perfect from the very first commit. Such approach prevents us from incremental development process and experimenting. It feels like adding `--experimental` flag would solve these issues. I'm pretty sure a lot of engineers would like to take risk and use experimental features. Thus, they would provide feedback and submit/fix issues that would be discovered. Basically, having this flag will let us to gather more feedback and run some sort of testing/experimenting on real-life code.
